spec.emu

```diff
@@ -15,37 +15,68 @@ contributors: Jordan Harband
 	<emu-alg>
 		1. Let _promise_ be the *this* value.
 		1. If IsPromise(_promise_) is *false*, throw a *TypeError* exception.
-		1. Let _thenFinally_ be ! CreateThenFinally(_onFinally_).
-		1. Let _catchFinally_ be ! CreateCatchFinally(_onFinally_).
+		1. Let _C_ be ? SpeciesConstructor(_promise_, %Promise%).
+		1. Let _thenFinally_ be ! CreateThenFinally(_C_, _onFinally_).
+		1. Let _catchFinally_ be ! CreateCatchFinally(_C_, _onFinally_).
 		1. Return ? Invoke(_promise_, `"then"`, &laquo; _thenFinally_, _catchFinally_ &raquo;).
 	</emu-alg>
 </emu-clause>
 
 <emu-clause id="sec-createthenfinally" aoid="CreateThenFinally">
-	<h1>CreateThenFinally ( _onFinally_ )</h1>
-	<p>The abstract operation CreateThenFinally takes an _onFinally_ function, and returns a callback function for use in Promise.prototype.finally.</p>
+	<h1>CreateThenFinally ( _C_, _onFinally_ )</h1>
+	<p>The abstract operation CreateThenFinally takes a Promise Constructor _C_, and an _onFinally_ function, and returns a callback function for use in Promise.prototype.finally.</p>
 	<emu-alg>
+		1. Assert: IsConstructor(_C_) is *true*.
 		1. If IsCallable(_onFinally_) is not *true*, return _onFinally_.
-		1. Return a function that takes one argument, _value_, and when invoked, performs the following steps:
-			1. Let _result_ be ? Call(_onFinally_, *undefined*).
-			1. Let _promise_ be ! PromiseResolve(%Promise%, _result_).
-			1. Let _valueThunk_ be equivalent to a function that returns _value_.
-			1. Let _promiseCapability_ be ! NewPromiseCapability(%Promise%).
-			1. Return PerformPromiseThen(_promise_, _valueThunk_, *undefined*, _promiseCapability_).
+		1. Let _thenFinally_ be a new built-in function object as defined in <emu-xref href="#sec-thenfinallyfunction">ThenFinally Function</emu-xref>.
+		1. Set _thenFinally_.[[Constructor]] to _C_.
+		1. Set _thenFinally_.[[onFinally]] to _onFinally_.
+		1. Return _thenFinally_.
+	</emu-alg>
+</emu-clause>
+
+<emu-clause id="sec-thenfinallyfunction" aoid="ThenFinallyFunction">
+	<h1>ThenFinally Function</h1>
+	<p>A ThenFinally function is an anonymous built-in function that has a [[Constructor]] and an [[onFinally]] internal slot. The value of the [[Constructor]] internal slot is a Promise constructor function object, and the value of the [[onFinally]] internal slot is a function object.</p>
+	<p>When a ThenFinally function _F_ is called with argument _value_, the following steps are taken:</p>
+	<emu-alg>
+		1. Let _onFinally_ be _F_.[[onFinally]].
+		1. Assert: IsCallable(_onFinally_) is *true*.
+		1. Let _result_ be ? Call(_onFinally_, *undefined*).
+		1. Let _C_ be _F_.[[Constructor]].
+		1. Assert: IsConstructor(_C_) is *true*.
+		1. Let _promise_ be ! PromiseResolve(_C_, _result_).
+		1. Let _valueThunk_ be equivalent to a function that returns _value_.
+		1. Return ? Invoke(_promise_, `"then"`, &laquo; _valueThunk_, *undefined* &raquo;).
```

Comment:
It will be a good test262 test that `undefined` is passed to subclasses (i.e. `arguments.length === 2`) :)